### PR TITLE
Spark Local Run job integration test with mocked DFS

### PR DESCRIPTION
**Pull request by wezhang**

### CODE REVIEW

**Comment by wezhang** [comment]
`Running com.microsoft.azure.hdinsight.spark.mock.SparkLocalRunnerIntegrationTest
Error: Could not find or load main class `

There is an issue that can't find the testing Spark Job class in classpath.

File: Utils/hdinsight-node-common/pom.xml [hunk 1]
```diff
@@ -255,5 +320,88 @@
...
+                <exclusion>
+                    <groupId>com.fasterxml.jackson.core</groupId>
+                    <artifactId>jackson-module-scala_2.11</artifactId>
+                </exclusion>
+            </exclusions>
+        </dependency>
+        <dependency>
+            <groupId>xml-apis</groupId>
+            <artifactId>xml-apis</artifactId>
+            <version>1.4.01</version>
```
Comment (voidfunction) [inline Utils/hdinsight-node-common/pom.xml]: version should be defined in Utils/pom.xml

File: Utils/hdinsight-node-common/pom.xml [hunk 2]
```diff
@@ -255,5 +320,88 @@
...
+                <exclusion>
+                    <groupId>com.fasterxml.jackson.core</groupId>
+                    <artifactId>jackson-module-scala_2.11</artifactId>
+                </exclusion>
+            </exclusions>
+        </dependency>
+        <dependency>
+            <groupId>org.apache.spark</groupId>
+            <artifactId>spark-sql_2.11</artifactId>
+            <version>2.1.0</version>
```
Comment (voidfunction) [inline Utils/hdinsight-node-common/pom.xml]: version issue.

File: Utils/hdinsight-node-common/pom.xml [hunk 3]
```diff
@@ -255,5 +320,88 @@
             <artifactId>kotlin-test</artifactId>
             <scope>test</scope>
         </dependency>
+        <dependency>
+            <groupId>org.assertj</groupId>
+            <artifactId>assertj-core</artifactId>
+            <!-- use 2.8.0 for Java 7 projects -->
+            <version>3.8.0</version>
```
Comment (voidfunction) [inline Utils/hdinsight-node-common/pom.xml]: version.

File: Utils/hdinsight-node-common/pom.xml [hunk 4]
```diff
@@ -125,6 +175,21 @@
...
+                <version>3.3.1</version>
+                <executions>
+                    <execution>
+                        <goals>
+                            <goal>testCompile</goal>
+                        </goals>
+                    </execution>
+                </executions>
+                <configuration>
+                    <scalaVersion>2.11.11</scalaVersion>
```
Comment (voidfunction) [inline Utils/hdinsight-node-common/pom.xml]: define a property in pom file?

File: Utils/hdinsight-node-common/pom.xml [hunk 5]
```diff
@@ -81,6 +82,55 @@
...
+                            }
+                        ]]></script>
+                    </scripts>
+                </configuration>
+                <dependencies>
+                    <dependency>
+                        <groupId>org.codehaus.groovy</groupId>
+                        <artifactId>groovy-all</artifactId>
+                        <!-- any version of Groovy \>= 1.5.0 should work here -->
+                        <version>2.4.12</version>
```
Comment (voidfunction) [inline Utils/hdinsight-node-common/pom.xml]: version? why set scope to runtime?

File: Utils/hdinsight-node-common/Test/scala/com/microsoft/azure/hdinsight/spark/mock/jobapp/WordCountTest.scala [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+    val outputFileOption = if (args.length <= 1) None else Some(args(1))
+    val conf = new SparkConf().setAppName("wordCount")
+    // Create a Scala Spark Context.
+    val sc = new SparkContext(conf)
+    // Load our input data.
+    val input =  sc.textFile(inputFile)
+    // Split up into words.
+    val words = input.flatMap(line => line.split(" "))
+    // Transform into word and count.
+    val counts = words.map(word => (word, 1)).reduceByKey{case (x, y) => x + y}
```
Comment (voidfunction) [inline Utils/hdinsight-node-common/Test/scala/com/microsoft/azure/hdinsight/spark/mock/jobapp/WordCountTest.scala]: >reduceByKey{case (x, y) => x + y} [](start = 46, length = 33)

reduceByKey(_+_) ?

File: Utils/hdinsight-node-common/pom.xml (hunk 4 above)
Comment (wezhang) [reply Utils/hdinsight-node-common/pom.xml]: fix it.

File: Utils/hdinsight-node-common/pom.xml (hunk 3 above)
Comment (wezhang) [reply Utils/hdinsight-node-common/pom.xml]: fix it

File: Utils/hdinsight-node-common/pom.xml (hunk 2 above)
Comment (wezhang) [reply Utils/hdinsight-node-common/pom.xml]: I would like to keep these Spark related version and won't spread to the whole project. In fact, those Spark Job code should be separated into an independent project.

File: Utils/hdinsight-node-common/pom.xml (hunk 1 above)
Comment (wezhang) [reply Utils/hdinsight-node-common/pom.xml]: fix it.

File: Utils/hdinsight-node-common/pom.xml [hunk 7]
```diff
@@ -125,6 +175,21 @@
                     <target>1.8</target>
                 </configuration>
             </plugin>
+            <plugin>
+                <groupId>net.alchim31.maven</groupId>
+                <artifactId>scala-maven-plugin</artifactId>
+                <version>3.3.1</version>
```
[the remark this answers is not in the document]
Comment (wezhang) [reply Utils/hdinsight-node-common/pom.xml]: Currently, all maven plugin version comes with the instance, I prefer to keep it here

File: Utils/hdinsight-node-common/pom.xml (hunk 5 above)
Comment (wezhang) [reply Utils/hdinsight-node-common/pom.xml]: Fix it with moving the script out.

File: Utils/hdinsight-node-common/Test/scala/com/microsoft/azure/hdinsight/spark/mock/jobapp/WordCountTest.scala (hunk 6 above)
Comment (wezhang) [reply Utils/hdinsight-node-common/Test/scala/com/microsoft/azure/hdinsight/spark/mock/jobapp/WordCountTest.scala]: Those codes are just copied from Spark Samples. :)

**Comment by wezhang** [comment]
```
< |    |         +--- com.microsoft.azure:azure-client-runtime:[1.1.0,2.0.0) -> 1.1.1 (*)
---
> |    |         +--- com.microsoft.azure:azure-client-runtime:[1.1.0,2.0.0) -> 1.2.0 (*)
+ echo ALERT: New dependencies introduced!
ALERT: New dependencies introduced!
```

**Comment by adashen** [comment]
@wezhang Please manage the version of azure-client-runtime and azure-client-authentication in the gradle file to version 1.1.1. In eclipse, we managed these version in parent pom. However in intelliJ, we didn't do so. Since we are currently pause the feature development. It seems there is no need to upgrade the version.

**Comment by wezhang** [comment]
```
+ cat /home/jenkins/dependencyLog//patch/PR-1250.patch
+ wc -l
+ [ 43 -eq 0 ]
+ cat /home/jenkins/dependencyLog//patch/PR-1250.patch
40a41,60
> +--- com.microsoft.azure:azure-client-runtime:1.1.1 (*)
> +--- com.microsoft.azure:azure-client-authentication:1.1.1
> |    +--- com.microsoft.azure:azure-client-runtime:1.1.1 (*)
> |    +--- com.microsoft.azure:adal4j:1.1.2
> |    |    +--- com.nimbusds:oauth2-oidc-sdk:4.5
> |    |    |    +--- javax.mail:mail:1.4.7
> |    |    |    |    \--- javax.activation:activation:1.1
> |    |    |    +--- net.jcip:jcip-annotations:1.0
> |    |    |    +--- org.apache.commons:commons-lang3:3.3.1 -> 3.5
> |    |    |    +--- commons-codec:commons-codec:1.9 -> 1.10
> |    |    |    +--- net.minidev:json-smart:1.1.1
> |    |    |    +--- com.nimbusds:lang-tag:1.4
> |    |    |    |    \--- net.minidev:json-smart:1.1.1
> |    |    |    \--- com.nimbusds:nimbus-jose-jwt:3.1.2
> |    |    |         +--- net.jcip:jcip-annotations:1.0
> |    |    |         +--- net.minidev:json-smart:1.1.1
> |    |    |         \--- org.bouncycastle:bcprov-jdk15on:1.51 -> 1.52
> |    |    +--- com.google.code.gson:gson:2.2.4 -> 2.7
> |    |    \--- commons-codec:commons-codec:1.10
> |    \--- com.microsoft.azure:azure-annotations:1.2.0
45,63c65
< |    |    +--- com.microsoft.azure:azure-client-authentication:[1.1.0,2.0.0) -> 1.1.1
< |    |    |    +--- com.microsoft.azure:azure-client-runtime:1.1.1 (*)
< |    |    |    +--- com.microsoft.azure:adal4j:1.1.2
< |    |    |    |    +--- com.nimbusds:oauth2-oidc-sdk:4.5
< |    |    |    |    |    +--- javax.mail:mail:1.4.7
< |    |    |    |    |    |    \--- javax.activation:activation:1.1
< |    |    |    |    |    +--- net.jcip:jcip-annotations:1.0
< |    |    |    |    |    +--- org.apache.commons:commons-lang3:3.3.1 -> 3.5
< |    |    |    |    |    +--- commons-codec:commons-codec:1.9 -> 1.10
< |    |    |    |    |    +--- net.minidev:json-smart:1.1.1
< |    |    |    |    |    +--- com.nimbusds:lang-tag:1.4
< |    |    |    |    |    |    \--- net.minidev:json-smart:1.1.1
< |    |    |    |    |    \--- com.nimbusds:nimbus-jose-jwt:3.1.2
< |    |    |    |    |         +--- net.jcip:jcip-annotations:1.0
< |    |    |    |    |         +--- net.minidev:json-smart:1.1.1
< |    |    |    |    |         \--- org.bouncycastle:bcprov-jdk15on:1.51 -> 1.52
< |    |    |    |    +--- com.google.code.gson:gson:2.2.4 -> 2.7
< |    |    |    |    \--- commons-codec:commons-codec:1.10
< |    |    |    \--- com.microsoft.azure:azure-annotations:1.2.0
---
> |    |    +--- com.microsoft.azure:azure-client-authentication:[1.1.0,2.0.0) -> 1.1.1 (*)
+ echo ALERT: New dependencies introduced!
ALERT: New dependencies introduced!
```

File: Utils/hdinsight-node-common/build-tools/cloneWinutils.kts [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+
+if (System.getProperty("os.name").toLowerCase().contains("windows")) {
+    val winutilsDir = File(Paths.get(targetDir, "tools", "winutils").toString())
+
+    if (winutilsDir.exists()) {
+        log.info("Winutils has been cloned to $winutilsDir, skip clone.")
+    } else {
+        log.info("Clone winutils from https://github.com/steveloughran/winutils to $winutilsDir")
+
+        val proc = ProcessBuilder("git", "clone", "--depth", "1", "https://github.com/steveloughran/winutils", "$winutilsDir")
```
Comment (voidfunction) [inline Utils/hdinsight-node-common/build-tools/cloneWinutils.kts]: >ProcessBuilder("git", "clone", "--depth", "1", "https://github.com/steveloughran/winutils", "$winutilsDir") [](start = 19, length = 107)

clone from a specific branch or commit id? master branch may have a build break when we fetch it.

Comment (wezhang) [reply Utils/hdinsight-node-common/build-tools/cloneWinutils.kts]: fix it.

**Comment by wezhang** [comment]
would fix #1210, #1224

**Pull request merged**